File: static/js/ChessPieces/Queen.js
class Queen extends ChessPiece{
	
	constructor(){
		super();
	}

	movablePlaces(x, y){
		return this.shrinkContinuosArray(this.getQueenMovablePlaces(x, y));
	}

	getQueenMovablePlaces(x, y){
		var rookPlaces = this.getRookMovablePlaces(x, y);
		var bishopPlaces = this.getBishopMovablePlaces(x, y);
		var place_ids = rookPlaces.concat(bishopPlaces);
		return place_ids;
	}

	getBishopMovablePlaces(x, y){
		var matrix = this.live_chessboard_matrix;
		var placeIds = [];
		var leftUp = [];
		var rightUp = [];
		var leftDown = [];
		var rightDown = [];
		if(x>=0&&x<=7&&y>=0&&y<=7){
			for(var i=0; i<=7; i++){
				var b_x = x - i;
				var b_y = y - i;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x<x&&b_y<y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							leftUp[i] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
			for(var j=0; j<=7; j++){
				var b_x = x + j;
				var b_y = y - j;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x>x&&b_y<y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							rightUp[j] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
			for(var k=0; k<=7; k++){
				var b_x = x - k;
				var b_y = y + k;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x<x&&b_y>y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							leftDown[k] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
			for(var n=0; n<=7; n++){
				var b_x = x + n;
				var b_y = y + n;
				if(b_x<=7&&b_x>=0&&b_y<=7&&b_y>=0){
					if(b_x>x&&b_y>y){
						var nextElement = document.getElementById(this.id_gen(b_y, b_x));
						if(nextElement!=null){
							rightDown[n] = nextElement.id;
							if(nextElement.childElementCount!=0){
								break;
							}
						}
					}
				}
			}
		}
		leftUp = this.moveArrayToBack(leftUp);
		rightUp = this.moveArrayToBack(rightUp);
		leftDown = this.moveArrayToBack(leftDown);
		rightDown = this.moveArrayToBack(rightDown);
		var leftToRight = leftDown.concat(rightUp);
		var rightToLeft = rightDown.concat(leftUp);
		placeIds = leftToRight.concat(rightToLeft);
		return placeIds;
	}

	getRookMovablePlaces(x, y){
		var matrix = this.live_chessboard_matrix;
		var placeIds = [];
		var left = [];
		var right = [];
		var up = [];
		var down = [];
		var leftCounter = 0;
		var upCounter = 0;

		if(x>=0&&x<=7&&y>=0&&y<=7){

			//Going Left
			for(var i=7; i>=0; i--){
				if(i<x){
					// console.log("hello y, x and i: "+y+" "+x+" "+i);
					var nextElement = document.getElementById(this.id_gen(y, i));
					if(nextElement!=null){
						left[leftCounter] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
						leftCounter++;
					}
				}
			}

			//Going Right
			for(var j=0; j<=7; j++){
				if(j>x){
					var nextElement = document.getElementById(this.id_gen(y, j));
					if(nextElement!=null){
						right[j] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
					}
				}
			}

			//Going Down
			for(var k=0; k<=7; k++){
				if(k>y){
					var nextElement = document.getElementById(this.id_gen(k, x));
					if(nextElement!=null){
						down[k] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
					}
				}
			}

			//Going Up
			for(var n=7; n>=0; n--){
				if(n<y){
					var nextElement = document.getElementById(this.id_gen(n, x));
					if(nextElement!=null){
						up[upCounter] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
						upCounter++;
					}
				}
			}	
		}
		
		left = this.moveArrayToBack(left);
		right = this.moveArrayToBack(right);
		up = this.moveArrayToBack(up);
		down = this.moveArrayToBack(down);

		var horizontal = left.concat(right);
		var vertical = up.concat(down);
		placeIds = horizontal.concat(vertical);
		// console.log("Rook Place Ids: "+placeIds)
	
		return placeIds;
	}
}

File: chess_app/ChessEngine/ChessPieces/Queen.py
from ChessPiece import ChessPiece
from chess_app.ChessEngine.Selecter import Select 

class Queen(ChessPiece):
	
	def __init__(self):
		super(Queen, self).__init__()

	def movablePlaces(self, x, y):
		return self.shrinkContinuosArray(self.getQueenMovablePlaces(x, y))

	def getQueenMovablePlaces(self, x, y):
		rookPlaces = self.getRookMovablePlaces(x, y)
		bishopPlaces = self.getBishopMovablePlaces(x, y)
		place_ids = rookPlaces + bishopPlaces
		
		return place_ids

	def getBishopMovablePlaces(self, x, y):
		matrix = self.live_chessboard_matrix
		placeIds = []
		leftUp = ["" for a in range(8)]
		rightUp = ["" for b in range(8)]
		leftDown = ["" for c in range(8)]
		rightDown = ["" for d in range(8)]
		
		if x >= 0 and x <= 7 and y >= 0 and y <= 7:
			
			for i in range(8):
				b_x = x - i
				b_y = y - i
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x < x and b_y < y:
						rightDownSelect = Select()
						nextElement = rightDownSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id != None:
							rightDown[i] = nextElement.parent_id
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
									
					
			for j in range(8):
				b_x = x + j
				b_y = y - j
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x > x and b_y < y:
						leftDownSelect = Select()
						nextElement = leftDownSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id != None:
							leftDown[j] = nextElement.parent_id
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
							
			for k in range(8):
				b_x = x - k
				b_y = y + k
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x < x and b_y > y:
						rightUpSelect = Select()
						nextElement = rightUpSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id!= None:
							rightUp[k] = nextElement.parent_id
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break						
					
			for n in range(8):
				b_x = x + n
				b_y = y + n
				
				if b_x <= 7 and b_x >= 0 and b_y <= 7 and b_y >= 0:
					if b_x > x and b_y > y:
						leftUpSelect = Select()
						nextElement = leftUpSelect.selectFromParentId(matrix, self.id_gen(b_y, b_x))
						
						if nextElement.parent_id!= None:
							leftUp[n] = nextElement.parent_id
							if nextElement.piece_id != "" and nextElement.piece_id != None:
								break
					
		leftUp = self.moveArrayToBack(leftUp)
		rightUp = self.moveArrayToBack(rightUp)
		leftDown = self.moveArrayToBack(leftDown)
		rightDown = self.moveArrayToBack(rightDown)
		
		leftToRight = leftDown + rightUp
		rightToLeft = rightDown + leftUp
		placeIds = leftToRight + rightToLeft
		
		return placeIds
	
	def getRookMovablePlaces(self, x, y):
		matrix = self.live_chessboard_matrix
		placeIds = []
		left = ["" for a in range(8)]
		right = ["" for b in range(8)]
		up = ["" for c in range(8)]
		down = ["" for d in range(8)]
		rightCounter = 0
		downCounter = 0

		if x >= 0 and x <= 7 and y >= 0 and y <= 7:

			#Going Left
			for i in range(7, -1, -1):
				if i < x:
					rightSelect = Select()
					nextElement = rightSelect.selectFromParentId(matrix, self.id_gen(y, i))
					if nextElement.parent_id != None:
						right[rightCounter] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
						rightCounter = rightCounter + 1
					
	

			#Going Right
			for j in range(8):
				if j > x:
					leftSelect = Select()
					nextElement = leftSelect.selectFromParentId(matrix, self.id_gen(y, j))
					if nextElement.parent_id != None:
						left[j] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
					

			#Going Down
			for k in range(8):
				if k > y:
					upSelect = Select()
					nextElement = upSelect.selectFromParentId(matrix, self.id_gen(k, x))
					if nextElement.parent_id != None:
						up[k] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
					
					
				
			#Going Up
			for n in range(7, -1, -1):
				if n < y:
					downSelect = Select()
					nextElement = downSelect.selectFromParentId(matrix, self.id_gen(n, x))
					if nextElement.parent_id != None:
						down[downCounter] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
						downCounter = downCounter + 1
				
					
				
		left = self.moveArrayToBack(left)
		right = self.moveArrayToBack(right)
		up = self.moveArrayToBack(up)
		down = self.moveArrayToBack(down)

		horizontal = left + right
		vertical = up + down
		placeIds = horizontal + vertical
	
		return placeIds





	